src/librustc/util/ppaux.rs

```diff
@@ -329,23 +332,32 @@ impl PrintContext {
                     assert_eq!(has_self, false);
                 } else {
                     // Types and traits.
-                    num_regions = generics.regions.len();
-                    num_types = generics.types.len();
+                    param_counts = child_param_counts;
                 }
             }
 
             if !verbose {
-                if generics.types.last().map_or(false, |def| def.has_default) {
-                    if let Some(substs) = tcx.lift(&substs) {
-                        let tps = substs.types().rev().skip(child_types);
-                        for (def, actual) in generics.types.iter().rev().zip(tps) {
-                            if !def.has_default {
-                                break;
-                            }
-                            if tcx.type_of(def.def_id).subst(tcx, substs) != actual {
-                                break;
+                let mut type_params =
+                    generics.params.iter().rev().filter_map(|param| {
+                        match param.kind {
+                            GenericParamDefKind::Type(ty) => Some((param.def_id, ty.has_default)),
+                            GenericParamDefKind::Lifetime(_) => None,
+                        }
+                    });
+                if let Some(last_ty) = type_params.next() {
+                    let (_, has_default) = last_ty;
+                    if has_default {
```

Comment:
You can use `peekable()` then `peek()` here to avoid recomposing an iterator with `chain` below.